Ring buffer for shared informer notifications

ash2k opened the pull request:
**What this PR does / why we need it**:
Improves memory allocation for shared informer listeners. Instead of always appending to the slice use as a ring buffer, avoiding reslice operations as long as there is room in the slice. See https://github.com/kubernetes/kubernetes/pull/47045#issuecomment-317621259 for details. This is a follow up PR for #47045.

Results from BenchmarkListener:
```
Current code (from the #47045):
1000000	      1540 ns/op	     109 B/op	       1 allocs/op
```
```
New code:
1000000	      1162 ns/op	      16 B/op	       1 allocs/op
```

**Special notes for your reviewer**:
Only review the last commit, this branch is based on #47045 PR. I'll rebase onto master once it is merged.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```
/kind enhancement
/sig api-machinery
/cc @deads2k @ncdc

ash2k commented:
I understand that the benchmark in this PR is kind of a hack - it just pre-allocates enough memory to avoid allocations. But I don't see any obvious way to test the usual usage pattern. Any suggestions? I just wanted to show that this approach works in principle.

ncdc commented on staging/src/k8s.io/client-go/util/rbuf/growing.go:
```diff
@@ -0,0 +1,83 @@
+
+func (r *Growing) ReadOne() interface{} {
+	if r.readable == 0 {
+		panic(errors.New("no elements available in the buffer"))
```
Don't panic in library code

ash2k replied:
Well, it is a bug if `ReadOne()` is called when there are no items in the buffer - what should it do? I could change this method to return a `bool` along with the value. WDYT?
Other places in code do even nastier things - #49369.

ash2k replied:
Yes, I'll change it to return `bool`. I actually thought about it before submitting the PR but decided to get feedback first. Thanks.

ncdc commented:
Does the ring buffer need to be thread safe / should you comment explicitly that it's not?

caesarxuchao commented:
/assign @caesarxuchao @ncdc
/unassign @janetkuo @mwielgus

ncdc commented:
/assign @deads2k @timothysc

timothysc commented:
/cc @kubernetes/sig-scalability-pr-reviews  @wojtek-t @gmarek

ash2k commented:
Just to make it clear - I'll resume work on this PR after #47045 is merged.

spiffxp commented:
/retest

spiffxp commented:
/ok-to-test

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -465,6 +466,13 @@ type processorListener struct {
+	// pendingNotifications is an unbounded ring buffer that holds all notifications not yet distributed
+	// there is one per listener, but a failing/stalled listener will have infinite pendingNotifications
+	// added until we OOM.
+	// TODO This is no worse than before, since reflectors were backed by unbounded DeltaFIFOs, but
```
nit: s/TODO/TODO:/

wojtek-t commented on staging/src/k8s.io/client-go/util/rbuf/growing.go:
```diff
@@ -0,0 +1,83 @@
+limitations under the License.
+*/
+
+package rbuf
```
Can we change it to full name instead of this shortcut, which is hard to guess without knowing what it is?

wojtek-t replied:
Maybe even call the directory: "buffer" and this particular file ring_growing.go or sth like that?

wojtek-t commented on staging/src/k8s.io/client-go/util/rbuf/growing.go:
```diff
@@ -0,0 +1,83 @@
+)
+
+// Growing is a growing ring buffer.
+// Not goroutine safe.
```
s/goroutine/thread/

wojtek-t commented on staging/src/k8s.io/client-go/util/rbuf/growing.go:
```diff
@@ -0,0 +1,83 @@
+func (r *Growing) WriteOne(data interface{}) {
+	if r.readable == r.n {
+		// Time to grow
+		newN := r.n + r.growStep
```
I don't think growing it by constant number of elements is a good idea.

Imagine that you start with size N and want to grow it to N^2, with growStep = N. that means:
- to grow to n^2 to require n^2 WriteOne ops
- you will have to grow it n times
- on average cost growing will be O(n^2)

So to summarize, you an amortized time of a single operation will be O(n). Which doesn't sound good.

Why can't we just always multiple the size of buffer by 2 (which is what Go allocator is doing for append BTW)?

ash2k replied:
Sure. I just wanted to reduce the speed with which it grows if things are slowing down. But anyway, we should address the TODO and have a capacity limit after which it stops growing. The question is, of course, what to do with notifications in that case.

ash2k commented:
@wojtek-t comments addressed. Thanks for feedback!

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -465,6 +471,13 @@ type processorListener struct {
 
 	handler ResourceEventHandler
 
+	// pendingNotifications is an unbounded ring buffer that holds all notifications not yet distributed
```
nit: "." at the end of sentence. Capital letter in the next line.

ash2k replied:
It's not my comment, I just moved it :)

ash2k replied:
Done anyway

wojtek-t reviewed (commented):
Just one minor nit - other than that LGTM.

@ash2k - can you please run your benchmark again to see if nothing changed in the meantime.

@ncdc - do you also want to take a look?

ash2k commented:
Same result:
```
1000000	      1139 ns/op	      16 B/op	       1 allocs/op
```

wojtek-t commented:
@ash2k - thanks.

LGTM though I will let others to also take a look. @ncdc @caesarxuchao

ncdc commented:
LGTM as well. Anyone else want a look? @timothysc @smarterclayton @deads2k @sttts ?

deads2k commented:
I like it.

wojtek-t commented:
/retest

ash2k commented:
Those unit tests are flaking on master too on my laptop.

wojtek-t commented:
/retest

ash2k commented:
/retest

spiffxp commented:
/retest

ash2k commented:
Can someone approve/LGTM please?

ncdc commented:
/lgtm

ash2k commented:
/retest

ash2k committed:
Ring buffer for notifications

Results from BenchmarkListener:
Current code:
1000000	      1540 ns/op	     109 B/op	       1 allocs/op

New:
1000000	      1139 ns/op	      16 B/op	       1 allocs/op

ncdc commented:
@ash2k what was the new commit - did you rebase?

ash2k commented:
@ncdc Yes, my branch became incompatible with master because of some Godeps-related changes in master (but not a merge conflict). Build was failing but I couldn't reproduce the issue locally on my branch. So I rebased and then found the issue. Had to run `./hack/update-staging-godeps.sh` to fix it. All the changes to `Godeps.json` files are the result of this. I haven't touched code at all.

But now it seems other things explode :( Will investigate later. It is so annoying - 99.99% sure it is not because of my changes.

ash2k commented:
/retest

ash2k commented:
FYI `pull-kubernetes-unit` failed because of https://github.com/kubernetes/kubernetes/issues/53124 , other ones - I don't know.

ncdc commented:
/lgtm
/approve

ncdc commented:
@deads2k do you want an issue for this PR to "fix", or are you ok with no-issue?

deads2k commented:
> @deads2k do you want an issue for this PR to "fix", or are you ok with no-issue?

I don't care, but I don't have the power to approve it.

ncdc commented:
@smarterclayton you ok approving this?

smarterclayton commented:
/approve no-issue

On Fri, Sep 29, 2017 at 2:34 PM, Andy Goldstein <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> you ok approving this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/49705#issuecomment-333204486>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3Pe14XbS3i872ljeFjbQjLjBBl2ks5snThOgaJpZM4OlJGA>
> .
>

ash2k commented:
/kind feature

ash2k commented:
/kind cleanup
